design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,208 @@
+
+## User Stories
+As a cluster-admin, I want an automated way to update the machine type of all VMs to be compatible with RHEL 10 without interrupting my workflow.
+As a cluster-admin, I want to be able to include only certain VMs to be updated (e.g. by namespace).
+As a cluster-admin, I want to be in control of the behavior of running VMs being updated; I can determine whether I want to manually restart running VMs being updated or let the automation restart all of them for me. 
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command that will allow the user to automatically update the machine type of any VMs with a machine type that is no longer supported. This command invokes a k8s `Job` that iterates through all VMs within a specified namespace (or all namespaces if none is specified), determines if the machine type is no longer supported, and updates it to the latest supported version. 
+
+If the VM is running, the update will not take effect until it is restarted, so the label `restart-vm-required` will be applied to the VM until the VMI is removed (indicating the VM has been stopped), upon which the label will be cleared. The job will only terminate once all `restart-vm-required` labels have been cleared. By default, the user must handle stopping/restarting the labelled VMs manually. This is to allow the user to safely update the machine types of their VMs at a time they choose without worrying about workload interruptions. However, the user also has the option to allow the job to automatically restart every running VM that is being updated and apply the changes immediately.
+
+The options to select a specific namespace or restart the VMs immediately are configurable with the `NAMESPACE` and `RESTART_NOW` environment variables respectively.
+
+Given a VM with the following spec:
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel8.2.0
+````
+When the MMTT is invoked, the VM spec will be updated to: 
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel9.2.0
+````
+where `pc-q35-rhel9.2.0` is the latest machine type version.
+
+Along with machine types being in the format `pc-q35-rhelx.x.x`, a VM may have a machine type of `q35`, an alias for the current latest machine type version *at the time the VM is started*.
+
+For example, this VM:
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        machine:
+          type: q35
+````
+may have been started when a RHEL-8 machine type was the latest machine type version and not been stopped or restarted since. Its VMI would thus look like this:
```

Comment:
Odd, any idea why we mutate the `VMI` with this? QEMU should be able to track this across migrations so I can't think why we would need to track it on our side.

Reply:
I think you're both saying the same thing. migrations aren't restarts.

Reply:
As for why it needs to be tracked at the VMI level? Only virt-launcher has access to qemu directly.

Reply:
I'm asking specifically why we expand the QEMU alias of `q35` from the VM to `pc-q35-rhel8.2.0` in the VMI instead of just using the `q35` alias for both and passing that to QEMU?

Reply:
Ah! We don't mutate the `VirtualMachineInstance` but somehow a static type is listed in the domain XML?

```
$ ./cluster-up/kubectl.sh get vm/cirros -o json | jq .spec.template.spec.domain.machine.type
selecting podman as container runtime
"q35"

$ ./cluster-up/kubectl.sh get vmi/cirros -o json | jq .spec.domain.machine.type
selecting podman as container runtime
"q35"

$ ./cluster-up/kubectl.sh exec -ti virt-launcher-cirros-kjn2n bash
selecting podman as container runtime
[..]
$ virsh dumpxml 1 | grep q35
    <type arch='x86_64' machine='pc-q35-rhel9.2.0'>hvm</type>
```

Anyway labelling any `VirtualMachines` using an alias is still valid but the suggestion that the `VirtualMachineInstance` is mutated needs to be dropped.

Reply:
This does get reflected in the `status` of the `VirtualMachineInstance` FWIW:
```
./cluster-up/kubectl.sh get vmi/cirros -o json | jq .status.machine
selecting podman as container runtime
{
  "type": "pc-q35-rhel9.2.0"
}
```